{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cd C:\\Users\\ab42313\\OneDrive\\ABZ_PhD\\Programming\\NCBI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import generic_dna\n",
    "\n",
    "handle = open(\"genBankRecord_545778205.fasta\", \"rU\")\n",
    "fasta = list(SeqIO.parse(handle, \"fasta\", generic_dna))\n",
    "\n",
    "seqs = []\n",
    "for i in range(0,len(fasta)):\n",
    "    seqs.append(fasta[i].seq)\n",
    "    #print fasta[i].seq\n",
    "\n",
    "genome = seqs[0]\n",
    "\n",
    "#MiSeq primers (5'-3')\n",
    "#fw = 'TCGTCGGCAGCGTCAGATGTGTATAAGAGACAG'\n",
    "#rv = 'GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAG'\n",
    "\n",
    "#Primers from wiki, 8F and 1492R\n",
    "fw = 'AGAGTTTGATCCTGGCTCAG'\n",
    "rv = 'CGGTTACCTTGTTACGACTT'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "def simulate_pcr_run(genome, fw_primer, rv_primer, approximative_gene_lenght = 2600, mismatches = 2):\n",
    "    '''\n",
    "    -approximative_gene_lenght: the approximative expected lenght of the gene\n",
    "    -mismatches: is set on 2 as in an highly stringent PCR for MiSeq sequencing\n",
    "    '''\n",
    "    \n",
    "    \n",
    "    ### Hamming distance function to calculate mismatches\n",
    "    \n",
    "    \n",
    "    \n",
    "    def calculate_hamming_distance(sequence1,sequence2): ###If one is primer should be insered as sequence1\n",
    "    \n",
    "        sequence1 = sequence1.upper()              ### Differences must be only related to nucleotides\n",
    "        sequence2 = sequence2.upper()\n",
    "\n",
    "        if len(sequence1) >= len(sequence2):       ### Necessary not to go out of range\n",
    "            seq1 =sequence1\n",
    "            seq2 =sequence2\n",
    "        elif len(sequence1) < len(sequence2):\n",
    "            seq2 =sequence1\n",
    "            seq1 =sequence2\n",
    "\n",
    "        hamming = 0\n",
    "\n",
    "        nucleotides = ('A','C','G','T')\n",
    "        alt_nucleotides = ('W', 'S','M','K','R','Y','B','D','H','V')\n",
    "\n",
    "\n",
    "        for i in range(0, len(seq2)):\n",
    "\n",
    "            if seq1[i] in nucleotides:\n",
    "\n",
    "                if seq1[i] == seq2[i]:\n",
    "                    hamming += 0\n",
    "                    \n",
    "                continue\n",
    "            \n",
    "            elif seq1[i] in alt_nucleotides:\n",
    "\n",
    "\n",
    "                if seq1[i] == 'W':\n",
    "                    if seq2[i] != 'A' or seq2[i] != 'T':\n",
    "                        hamming += 1\n",
    "\n",
    "                if seq1[i] == 'S':\n",
    "                    if seq2[i] != 'C' or seq2[i] != 'G':\n",
    "                        hamming += 1\n",
    "\n",
    "                if seq1[i] == 'M':\n",
    "                    if seq2[i] != 'A' or seq2[i] != 'C':\n",
    "                        hamming += 1\n",
    "\n",
    "                if seq1[i] == 'K':\n",
    "                    if seq2[i] != 'G' or seq2[i] != 'T':\n",
    "                        hamming += 1\n",
    "\n",
    "                if seq1[i] == 'R':\n",
    "                    if seq2[i] != 'A' or seq2[i] != 'G':\n",
    "                        hamming += 1\n",
    "\n",
    "                if seq1[i] == 'Y':\n",
    "                    if seq2[i] != 'C' or seq2[i] != 'T':\n",
    "                        hamming += 1\n",
    "\n",
    "                if seq1[i] == 'B':\n",
    "                    if seq2[i] != 'C' or seq2[i] != 'G' or seq2[i] != 'T':\n",
    "                        hamming += 1\n",
    "                        \n",
    "                if seq1[i] == 'D':\n",
    "                    if seq2[i] != 'A' or seq2[i] != 'G' or seq2[i] != 'T':\n",
    "                        hamming += 1\n",
    "\n",
    "                if seq1[i] == 'H':\n",
    "                    if seq2[i] != 'A' or seq2[i] != 'C' or seq2[i] != 'T':\n",
    "                        hamming += 1\n",
    "\n",
    "                if seq1[i] == 'V':\n",
    "                    if seq2[i] != 'A' or seq2[i] != 'C' or seq2[i] != 'G':\n",
    "                        hamming += 1\n",
    "\n",
    "                if seq1[i] == 'N' or seq1[i] == '-':\n",
    "                    if seq2[i] != 'A' or seq2[i] != 'C' or seq2[i] != 'G'or seq2[i] != 'T':\n",
    "                        continue\n",
    "\n",
    "\n",
    "        if len(seq1) > len(seq2):                               ### Correcting the Hamming score for missing nucleotides\n",
    "\n",
    "            hamming += len(seq1) - len(seq2)\n",
    "\n",
    "        elif len(seq1) < len(seq2):\n",
    "\n",
    "            hamming += len(seq2) - len(seq1)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "        return hamming\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    ### PCR simulation function:\n",
    "    \n",
    "    \n",
    "    \n",
    "    #primer_positions = []\n",
    "    sequences = []\n",
    "    gene_len = approximative_gene_lenght + 100\n",
    "    fw = fw_primer\n",
    "    rv = rv_primer\n",
    "\n",
    "    positions = range(0,len(genome)-len(fw))\n",
    "    len(positions)\n",
    "\n",
    "    for i in positions:\n",
    "        \n",
    "        testing = genome[i:i+len(fw)]\n",
    "        \n",
    "        if calculate_hamming_distance(fw, testing) <= mismatches:\n",
    "            \n",
    "            for m in range(i + len(rv), gene_len):\n",
    "                \n",
    "                if calculate_hamming_distance(rv, genome[m : m + len(rv)]) <= mismatches:\n",
    "                    \n",
    "                    if len(genome[i:m+len(rv)]) > gene_len-150:         #''Discarting short reads\n",
    "                        \n",
    "                        #print i, m, genome[i:m]\n",
    "                        sequences.append(str(genome[i:m+len(rv)]))\n",
    "                        #primer_positions.appen+d(i)\n",
    "                        #primer_positions.append(m+len(rv))\n",
    "                  \n",
    "                        break\n",
    "    \n",
    "    return sequences\n",
    "    #return 'Positions: ', primer_positions\n",
    "    #return sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The program apparently works since it selects fragments.\n",
    "\n",
    "Problems:\n",
    "    - \"amplified\" fragments blast result is not informative at all (raises an error...)\n",
    "    - MiSeq primers don't work.\n",
    "    - Primer/Probe mismatches have to be set at high values\n",
    "    \n",
    "Will check the code again on monday."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
