{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C:\\Users\\ab42313\\Google Drive\\Research life\\PhD\\Projects\\In_silico_metagenomics\\data\n"
     ]
    }
   ],
   "source": [
    "cd C:\\Users\\ab42313\\Google Drive\\Research life\\PhD\\Projects\\In_silico_metagenomics\\data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import generic_dna\n",
    "\n",
    "handle = open(\"genBankRecord_545778205.fasta\", \"rU\")\n",
    "fasta = list(SeqIO.parse(handle, \"fasta\", generic_dna))\n",
    "\n",
    "seqs = []\n",
    "for i in range(0,len(fasta)):\n",
    "    seqs.append(fasta[i].seq)\n",
    "    #print fasta[i].seq\n",
    "\n",
    "genome = seqs[0]\n",
    "\n",
    "#MiSeq primers (5'-3')\n",
    "#fw = 'TCGTCGGCAGCGTCAGATGTGTATAAGAGACAG'\n",
    "#rv = 'GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAG'\n",
    "\n",
    "#Primers from wiki, 8F and 1492R\n",
    "fw = 'AGAGTTTGATCCTGGCTCAG'\n",
    "rv = 'CGGTTACCTTGTTACGACTT'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### PCR simulation function:\n",
    "\n",
    "\n",
    "def simulate_pcr_run(genome, fw_primer, rv_primer, approximative_gene_lenght = 2600, mismatches = 2):\n",
    "    '''\n",
    "    -approximative_gene_lenght: the approximative expected lenght of the gene\n",
    "    -mismatches: is set on 2 as in an highly stringent PCR for MiSeq sequencing\n",
    "    '''\n",
    "    \n",
    "    \n",
    "    ### Hamming distance function to calculate mismatches\n",
    "    \n",
    "\n",
    "\n",
    "\n",
    "    def calculate_hamming_distance(sequence1,sequence2): ###If one is a primer should be insered as sequence1\n",
    "\n",
    "        sequence1 = sequence1.upper()              ### Differences must be only related to nucleotides not to typing\n",
    "        sequence2 = sequence2.upper()\n",
    "\n",
    "\n",
    "        m = 0\n",
    "\n",
    "        if len(sequence1) >= len(sequence2):       ### Necessary not to go out of range\n",
    "            m = len(sequence1) - len(sequence2)\n",
    "            seq2 =sequence1[0:len(sequence2)]\n",
    "            seq1 =sequence2\n",
    "\n",
    "\n",
    "        else:\n",
    "            m = len(sequence2) - len(sequence1)\n",
    "            seq1 =sequence1\n",
    "            seq2 =sequence2[0:len(sequence1)]\n",
    "\n",
    "\n",
    "        hamming = m\n",
    "\n",
    "\n",
    "\n",
    "        nucleotides = ('A','C','G','T')\n",
    "        alt_nucleotides = ('W', 'S','M','K','R','Y','B','D','H','V')\n",
    "\n",
    "\n",
    "\n",
    "        for i in range(0, len(seq1)):\n",
    "\n",
    "\n",
    "            if seq1[i] == seq2[i]:\n",
    "                continue\n",
    "\n",
    "            if seq1[i] != seq2[i]:\n",
    "\n",
    "                if seq1[i] in nucleotides and seq2[i] in nucleotides:\n",
    "                    if seq1[i] != seq2[i]:\n",
    "                        hamming += 1\n",
    "\n",
    "                else:\n",
    "\n",
    "                    if seq1[i] == 'W':\n",
    "                        if seq2[i] != 'A' or seq2[i] != 'T':\n",
    "                            hamming += 1\n",
    "\n",
    "                    if seq1[i] == 'S':\n",
    "                        if seq2[i] != 'C' or seq2[i] != 'G':\n",
    "                            hamming += 1\n",
    "\n",
    "                    if seq1[i] == 'M':\n",
    "                        if seq2[i] != 'A' or seq2[i] != 'C':\n",
    "                            hamming += 1\n",
    "\n",
    "                    if seq1[i] == 'K':\n",
    "                        if seq2[i] != 'G' or seq2[i] != 'T':\n",
    "                            hamming += 1\n",
    "\n",
    "                    if seq1[i] == 'R':\n",
    "                        if seq2[i] != 'A' or seq2[i] != 'G':\n",
    "                            hamming += 1\n",
    "\n",
    "                    if seq1[i] == 'Y':\n",
    "                        if seq2[i] != 'C' or seq2[i] != 'T':\n",
    "                            hamming += 1\n",
    "\n",
    "                    if seq1[i] == 'B':\n",
    "                        if seq2[i] != 'C' or seq2[i] != 'G' or seq2[i] != 'T':\n",
    "                            hamming += 1\n",
    "\n",
    "                    if seq1[i] == 'D':\n",
    "                        if seq2[i] != 'A' or seq2[i] != 'G' or seq2[i] != 'T':\n",
    "                            hamming += 1\n",
    "\n",
    "                    if seq1[i] == 'H':\n",
    "                        if seq2[i] != 'A' or seq2[i] != 'C' or seq2[i] != 'T':\n",
    "                            hamming += 1\n",
    "\n",
    "                    if seq1[i] == 'V':\n",
    "                        if seq2[i] != 'A' or seq2[i] != 'C' or seq2[i] != 'G':\n",
    "                            hamming += 1\n",
    "\n",
    "                    if seq1[i] == 'N' or seq1[i] == '-':\n",
    "                        if seq2[i] != 'A' or seq2[i] != 'C' or seq2[i] != 'G'or seq2[i] != 'T':\n",
    "                            continue\n",
    "\n",
    "                    else:\n",
    "                        continue\n",
    "\n",
    "\n",
    "\n",
    "        return hamming\n",
    "\n",
    "    \n",
    "    sequences = []\n",
    "    gene_len = approximative_gene_lenght + 100\n",
    "    fw = fw_primer\n",
    "    rv = rv_primer\n",
    "    \n",
    "    Hint = mismatches + 1               ### H interval, refers to the possible mismetches allowed\n",
    "    \n",
    "    positions = range(0,len(genome))\n",
    "\n",
    "\n",
    "    for i in positions:\n",
    "\n",
    "        testing_fw = genome[i:i+len(fw)]\n",
    "        Hfw = calculate_hamming_distance(fw, testing_fw)\n",
    "        \n",
    "        \n",
    "        if Hfw > mismatches:\n",
    "            \n",
    "            i += Hfw - Hint\n",
    "\n",
    "\n",
    "        if Hfw <= mismatches:\n",
    "\n",
    "\n",
    "            #for m in range(i+1,len(genome)):\n",
    "            for m in range(i+1, i + gene_len):\n",
    "                \n",
    "                Hrv = calculate_hamming_distance(rv, genome[m:m+len(rv)])\n",
    "                \n",
    "                if Hrv > mismatches:\n",
    "            \n",
    "                    m += Hfw - Hint\n",
    "\n",
    "                if Hrv <= mismatches:\n",
    "\n",
    "                    sequences.append(genome[i + len(fw) : m])\n",
    "                    continue\n",
    "\n",
    "\n",
    "    return sequences\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
